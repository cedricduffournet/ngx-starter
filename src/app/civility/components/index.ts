export * from './civility-add/civility-add.component';
export * from './civility-update/civility-update.component';
export * from './civility-delete/civility-delete.component';
export * from './civility-form/civility-form.component';
export * from './civility-item/civility-item.component';
export * from './civility-items/civility-items.component';
