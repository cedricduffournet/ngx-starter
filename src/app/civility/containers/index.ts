export * from './civility-add-modal/civility-add-modal.component';
export * from './civility-update-modal/civility-update-modal.component';
export * from './civility-delete-modal/civility-delete-modal.component';
export * from './civility-list-view/civility-list-view.component';
