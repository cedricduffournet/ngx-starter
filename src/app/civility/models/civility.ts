import { schema } from 'normalizr';

export interface Civility {
  id: number;
  name: string;
  code: string;
}

export const civilitySchema = new schema.Entity('civilities');
