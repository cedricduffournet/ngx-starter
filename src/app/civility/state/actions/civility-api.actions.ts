import { props, createAction } from '@ngrx/store';

import { NormalizedData } from '@app/shared/models/normalized.model';

export const loadCivilitySuccess = createAction(
  '[Civility API] Load civility SUCCESS',
  props<{ civilities: NormalizedData }>()
);

export const loadCivilityFailure = createAction(
  '[Civility API] Load civility FAILURE',
  props<{ error: any }>()
);

export const addCivilitySuccess = createAction(
  '[Civility API] Add civility SUCCESS',
  props<{ civility: NormalizedData }>()
);

export const addCivilityFailure = createAction(
  '[Civility API] Add civility FAILURE',
  props<{ error: any }>()
);

export const updateCivilitySuccess = createAction(
  '[Civility/entities] Update civility SUCCESS',
  props<{ civility: NormalizedData }>()
);

export const updateCivilityFailure = createAction(
  '[Civility/entities] Update civility FAILURE',
  props<{ error: any }>()
);

export const deleteCivilitySuccess = createAction(
  '[Civility API] Remove civility SUCCESS',
  props<{ id: number }>()
);

export const deleteCivilityFailure = createAction(
  '[Civility API] Remove civility FAILURE',
  props<{ error: any }>()
);
