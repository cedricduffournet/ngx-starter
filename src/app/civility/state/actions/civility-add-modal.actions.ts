import { props, createAction } from '@ngrx/store';

import { Civility } from '@app/civility/models/civility';

export const addCivility = createAction(
  '[Add civility modal] Add civility',
  props<{ civility: Civility }>()
);
