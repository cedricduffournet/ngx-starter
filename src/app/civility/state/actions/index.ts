import * as CivilityUpdateModalActions from './civility-update-modal.actions';
import * as CivilityAddModalActions from './civility-add-modal.actions';
import * as CivilityDeleteModalActions from './civility-delete-modal.actions';
import * as CivilityListViewActions from './civility-list-view.actions';
import * as CivilityApiActions from './civility-api.actions';

export {
  CivilityListViewActions,
  CivilityApiActions,
  CivilityUpdateModalActions,
  CivilityAddModalActions,
  CivilityDeleteModalActions
};
