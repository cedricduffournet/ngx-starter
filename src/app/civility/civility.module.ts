import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { CivilityEffects } from '@app/civility/state/effects';
import {
  CivilityAddComponent,
  CivilityUpdateComponent,
  CivilityDeleteComponent,
  CivilityFormComponent,
  CivilityItemComponent,
  CivilityItemsComponent
} from '@app/civility/components';
import {
  CivilityAddModalComponent,
  CivilityUpdateModalComponent,
  CivilityDeleteModalComponent,
  CivilityListViewComponent
} from '@app/civility/containers';
import { reducers } from '@app/civility/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([CivilityEffects]),
    StoreModule.forFeature('civilities', reducers)
  ],
  declarations: [
    CivilityListViewComponent,
    CivilityAddComponent,
    CivilityUpdateComponent,
    CivilityDeleteComponent,
    CivilityFormComponent,
    CivilityItemComponent,
    CivilityItemsComponent,
    CivilityAddModalComponent,
    CivilityUpdateModalComponent,
    CivilityDeleteModalComponent
  ],
  entryComponents: [
    CivilityUpdateModalComponent,
    CivilityAddModalComponent,
    CivilityDeleteModalComponent
  ],
  exports: [CivilityListViewComponent]
})
export class CivilityModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
