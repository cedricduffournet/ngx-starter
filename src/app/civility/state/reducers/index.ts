import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers
} from '@ngrx/store';

import * as fromRoot from '@app/core/state/reducers';
import * as fromCivilityCollection from '@app/civility/state/reducers/civility-collection.reducer';
import * as fromCivilityEntities from '@app/civility/state/reducers/civility-entities.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { Civility } from '@app/civility/models/civility';
import { Authorization } from '@app/core/models/authorization.model';

export interface CivilitiesState {
  collection: fromCivilityCollection.State;
  civilities: fromCivilityEntities.State;
}

export interface State extends fromRoot.State {
  civilities: CivilitiesState;
}

export function reducers(state: CivilitiesState | undefined, action: Action) {
  return combineReducers({
    collection: fromCivilityCollection.reducer,
    civilities: fromCivilityEntities.reducer
  })(state, action);
}

export const selectCivilitiesState = createFeatureSelector<
  State,
  CivilitiesState
>('civilities');

export const getCivilityEntitiesState = createSelector(
  selectCivilitiesState,
  state => state.civilities
);

export const getCivilityEntities = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getEntities
);

export const getCivilityEntitiesUpdating = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getUpdating
);

export const getCivilityEntitiesUpdated = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getUpdated
);

export const getCivilityCollectionState = createSelector(
  selectCivilitiesState,
  state => state.collection
);
export const getCivilityIds = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getIds
);
export const getCivilityCollectionAdding = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getAdding
);
export const getCivilityCollectionAdded = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getAdded
);
export const getCivilityCollectionDeleting = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getDeleting
);
export const getCivilityCollectionDeleted = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getDeleted
);

export const getCivilities = createSelector(
  getCivilityEntities,
  getCivilityIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const getSelectedCivilityId = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getSelectedId
);

export const getSelectedCivility = createSelector(
  getCivilityEntities,
  getSelectedCivilityId,
  (civilityEntities, civilityId): Civility => {
    return civilityEntities[civilityId];
  }
);

export const canUpdateCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_EDIT']),
  canUpdate => canUpdate
);

export const canDeleteCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_DELETE']),
  canDelete => canDelete
);

export const canCreateCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_CREATE']),
  canCreate => canCreate
);

export const getCivilityAuthorization = createSelector(
  canUpdateCivility,
  canDeleteCivility,
  canCreateCivility,
  (canUpdate, canDelete, canAdd): Authorization => {
    return {
      update: canUpdate,
      delete: canDelete,
      create: canAdd
    };
  }
);
