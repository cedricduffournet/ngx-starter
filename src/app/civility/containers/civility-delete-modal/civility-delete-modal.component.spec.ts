import { TestBed, ComponentFixture } from '@angular/core/testing';

import { Store, MemoizedSelector } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { TranslateModule } from '@ngx-translate/core';

import { ModalWrapperModule } from '@app/shared/modal';
import { ValidationActionModule } from '@app/shared/validation-action';
import { CivilityDeleteComponent } from '@app/civility/components';
import { CivilityDeleteModalComponent } from '@app/civility/containers';
import * as fromCivilities from '@app/civility/state/reducers';
import { CivilityDeleteModalActions } from '@app/civility/state/actions';
import { Civility } from '@app/civility/models/civility';

describe('DeleteCivilityModalComponent', () => {
  let fixture: ComponentFixture<CivilityDeleteModalComponent>;
  let component: CivilityDeleteModalComponent;
  let store: MockStore<fromCivilities.State>;
  let deleted: MemoizedSelector<fromCivilities.State, boolean>;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [CivilityDeleteComponent, CivilityDeleteModalComponent],
      imports: [
        TranslateModule.forRoot(),
        ModalWrapperModule,
        ValidationActionModule
      ],
      providers: [provideMockStore(), BsModalRef]
    });

    fixture = TestBed.createComponent(CivilityDeleteModalComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    deleted = store.overrideSelector(
      fromCivilities.getCivilityCollectionDeleted,
      false
    );
    spyOn(store, 'dispatch');
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a deleteCivility event on submit', () => {
    const civility = {
      id: 1,
      code: 'code',
      name: 'name'
    } as Civility;
    const action = CivilityDeleteModalActions.deleteCivility({ civility });
    fixture.detectChanges();
    component.onDelete(civility);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should close modal after civility added', () => {
    spyOn(component.bsModalRef, 'hide');
    fixture.detectChanges();
    deleted.setResult(true);
    store.setState({} as any);
    expect(component.bsModalRef.hide).toHaveBeenCalled();
  });

  it('should close modal on cancel', () => {
    spyOn(component.bsModalRef, 'hide');
    component.onCancel();
    expect(component.bsModalRef.hide).toHaveBeenCalled();
  });
});
