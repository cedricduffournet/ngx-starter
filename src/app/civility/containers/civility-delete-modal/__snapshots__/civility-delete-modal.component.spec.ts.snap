// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteCivilityModalComponent should be created 1`] = `
<app-civility-delete-modal
  bsModalRef={[Function BsModalRef]}
  deleted$={[Function Observable]}
  deleting$="undefined"
  facade={[Function Object]}
  selectedCivility$="undefined"
  subscription={[Function Subscriber]}
>
  <app-modal-wrapper
    ng-reflect-title="CONFIRM"
  >
    <div
      class="modal-content bg-light"
    >
      
      <div
        class="modal-header"
      >
        <h4
          class="modal-title"
        >
          CONFIRM
        </h4>
      </div>
      <div
        class="modal-body p-0"
      >
        <app-civility-delete>
          <section
            class="bg-white"
          >
            <div
              class="p-3"
            >
               QUESTION_DELETE CIVILITY_DELETE 
            </div>
          </section>
          <app-wrapper-validation-action>
            <div
              class="d-flex justify-content-end align-items-center p-3"
            >
              <app-validation-button
                classconfirm="danger"
                ng-reflect-class-confirm="danger"
                ng-reflect-label-cancel="CANCEL"
                ng-reflect-label-confirm="DELETE"
              >
                <div
                  class="button-validation-action text-center d-flex justify-content-end align-items-center p-3-group"
                  ng-reflect-klass="button-validation-action text-"
                  ng-reflect-ng-class=""
                >
                  
                  <button
                    class="btn btn-cancel btn-lg btn-light"
                    ng-reflect-klass="btn btn-cancel btn-lg"
                    ng-reflect-ng-class="btn-light"
                    type="button"
                  >
                     CANCEL 
                  </button>
                  
                  <button
                    class="btn btn-confirm btn-lg btn-danger"
                    ng-reflect-klass="btn btn-confirm btn-lg"
                    ng-reflect-ng-class="btn-danger"
                    type="submit"
                  >
                    
                    
                    
                    <span>
                      DELETE
                    </span>
                  </button>
                </div>
              </app-validation-button>
            </div>
          </app-wrapper-validation-action>
        </app-civility-delete>
      </div>
    </div>
  </app-modal-wrapper>
</app-civility-delete-modal>
`;
